Client-side hashing fails on files >2GB in size, preventing upload.

This is being reopened, as it is a genuine bug. My additional comments:

This is an actual bug, not related to your reverse proxy config. The hashing fails, but the `crypto` module is not available if the browser does not consider the origin secure - as a result, the hashing is skipped altogether on upload when you are connecting directly. If your browser DOES consider the origin secure with your direct connection, it too will fail. 

In Chrome, you can test via defining the local hostname/IP in `chrome://flags/#unsafely-treat-insecure-origin-as-secure` and relaunching. Check if the necessary crypto module is available by running `crypto?.subtle?.digest` in the dev console after loading the page. If your connection is insecure, the value will be `undefined`. If it is secure, it should return `ƒ digest() { [native code] }`. 

[This is the conditional](https://github.com/immich-app/immich/blob/92412ca2f70bd117b44b5c8c6d19a68d3ed8e03e/web/src/lib/utils/file-uploader.ts#L138) that skips hashing if the crypto module is not available. Overriding the conditional to always return false allows the files >2GB in size to upload without issue on a secure origin. 

Reopening this as an issue. 

Output from Firefox is:
`TypeError: SubtleCrypto.digest: ArrayBuffer branch of (ArrayBufferView or ArrayBuffer) can't be an ArrayBuffer or an ArrayBufferView larger than 2 GB`

Should probably use a stream instead for hash generation to get around this. 

### Discussed in https://github.com/immich-app/immich/discussions/11804

<div type='discussions-op-text'>

<sup>Originally posted by **TheOnlyWayUp** August 15, 2024</sup>
### The bug

I'm uploading a `.mov` file, it's a bit large (3.8GB).

I'm on v1.112.1, and it fails with a `DOMException`.
![image](https://github.com/user-attachments/assets/66d15544-e8e1-41fd-a4d8-9941760fcd4a) (logs attached below)

I've ensured the reverse proxy supports large files. It might be an issue with a size limit for `ArrayBuffer` on chrome (2GB).

### The OS that Immich Server is running on

Ubuntu 24.04 LTS x86_64

### Version of Immich Server

v1.112.1

### Version of Immich Mobile App

n/a

### Platform with the issue

- [ ] Server
- [X] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
#
# WARNING: Make sure to use the docker-compose.yml of the current release:
#
# https://github.com/immich-app/immich/releases/latest/download/docker-compose.yml
#
# The compose file on main may not be compatible with the latest release.
#

name: immich

services:
  immich-server:
    container_name: immich_server
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    networks:
      - default
      - reverse_proxy
    extends:
      file: hwaccel.transcoding.yml
      service: cpu # set to one of [nvenc, quicksync, rkmpp, vaapi, vaapi-wsl] for accelerated transcoding
    volumes:
      # Do not edit the next line. If you want to change the media storage location on your system, edit the value of UPLOAD_LOCATION in the .env file
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - /etc/localtime:/etc/localtime:ro
      - /root/Projects/Immich/external:/mnt/media:ro
    env_file:
      - .env
    expose:
      - 3001
    depends_on:
      - redis
      - database
    restart: always
    healthcheck:
      disable: false
    labels:
      caddy: <host>
      caddy.reverse_proxy: "{{upstreams 3001}}"
      caddy.request_body.max_size: 128GB

  immich-machine-learning:
    container_name: immich_machine_learning
    networks:
      - default
    # For hardware acceleration, add one of -[armnn, cuda, openvino] to the image tag.
    # Example tag: ${IMMICH_VERSION:-release}-cuda
    image: ghcr.io/immich-app/immich-machine-learning:${IMMICH_VERSION:-release}
    # extends: # uncomment this section for hardware acceleration - see https://immich.app/docs/features/ml-hardware-acceleration
    #   file: hwaccel.ml.yml
    #   service: cpu # set to one of [armnn, cuda, openvino, openvino-wsl] for accelerated inference - use the `-wsl` version for WSL2 where applicable
    volumes:
      - model-cache:/cache
    env_file:
      - .env
    restart: always
    healthcheck:
      disable: false

  redis:
    container_name: immich_redis
    networks:
      - default
    image: docker.io/redis:6.2-alpine@sha256:e3b17ba9479deec4b7d1eeec1548a253acc5374d68d3b27937fcfe4df8d18c7e
    healthcheck:
      test: redis-cli ping || exit 1
    restart: always

  database:
    container_name: immich_postgres
    networks:
      - default
    image: docker.io/tensorchord/pgvecto-rs:pg14-v0.2.0@sha256:90724186f0a3517cf6914295b5ab410db9ce23190a2d9d0b9dd6463e3fa298f0
    environment:
      POSTGRES_PASSWORD: ${DB_PASSWORD}
      POSTGRES_USER: ${DB_USERNAME}
      POSTGRES_DB: ${DB_DATABASE_NAME}
      POSTGRES_INITDB_ARGS: '--data-checksums'
    volumes:
      # Do not edit the next line. If you want to change the database storage location on your system, edit the value of DB_DATA_LOCATION in the .env file
      - ${DB_DATA_LOCATION}:/var/lib/postgresql/data
    healthcheck:
      test: pg_isready --dbname='${DB_DATABASE_NAME}' --username='${DB_USERNAME}' || exit 1; Chksum="$$(psql --dbname='${DB_DATABASE_NAME}' --username='${DB_USERNAME}' --tuples-only --no-align --command='SELECT COALESCE(SUM(checksum_failures), 0) FROM pg_stat_database')"; echo "checksum failure count is $$Chksum"; [ "$$Chksum" = '0' ] || exit 1
      interval: 5m
      start_interval: 30s
      start_period: 5m
    command: ["postgres", "-c", "shared_preload_libraries=vectors.so", "-c", 'search_path="$$user", public, vectors', "-c", "logging_collector=on", "-c", "max_wal_size=2GB", "-c", "shared_buffers=512MB", "-c", "wal_compression=on"]
    restart: always

volumes:
  model-cache:

networks:
  default:
  reverse_proxy:
    external: true
```


### Your .env content

```Shell
# You can find documentation for all the supported env variables at https://immich.app/docs/install/environment-variables

# The location where your uploaded files are stored
UPLOAD_LOCATION=<absolute_path>
# The location where your database files are stored
DB_DATA_LOCATION=<absolute_path>

# To set a timezone, uncomment the next line and change Etc/UTC to a TZ identifier from this list: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones#List
# TZ=Etc/UTC

# The Immich version to use. You can pin this to a specific version like "v1.71.0"
IMMICH_VERSION=release

# Connection secret for postgres. You should change it to a random password
DB_PASSWORD=<>

# The values below this line do not need to be changed
###################################################################################
DB_USERNAME=<>
DB_DATABASE_NAME=<>
```


### Reproduction steps

```bash
1. Open the Web UI
2. Upload an asset >2gb
```


### Relevant log output

```shell
Error with Permissions-Policy header: Origin trial controlled feature not enabled: 'run-ad-auction'.
Error with Permissions-Policy header: Origin trial controlled feature not enabled: 'join-ad-interest-group'.
Error with Permissions-Policy header: Origin trial controlled feature not enabled: 'browsing-topics'.
bootstrap-legacy-autofill-overlay.js:870 [Violation] 'requestIdleCallback' handler took 59ms
file-uploader.DcK3d__D.js:1 Error calculating sha1 file=aftermovie.mov) DOMException: The requested file could not be read, typically due to permission problems that have occurred after a reference to a file was acquired.
K @ file-uploader.DcK3d__D.js:1
await in K (async)
(anonymous) @ file-uploader.DcK3d__D.js:1
(anonymous) @ file-uploader.DcK3d__D.js:1
tryRun @ file-uploader.DcK3d__D.js:1
(anonymous) @ file-uploader.DcK3d__D.js:1
addTask @ file-uploader.DcK3d__D.js:1
W @ file-uploader.DcK3d__D.js:1
await in W (async)
m @ 2.DdyS7M7E.js:1
v @ 2.DdyS7M7E.js:1
(anonymous) @ scheduler.D4cMnZ5O.js:1
(anonymous) @ scheduler.D4cMnZ5O.js:1
handle-error.CpLOUvPZ.js:1 [handleError]: Unable to upload file ProgressEvent {isTrusted: true, lengthComputable: false, loaded: 0, total: 0, type: 'error', …} undefined
E @ handle-error.CpLOUvPZ.js:1
K @ file-uploader.DcK3d__D.js:1
await in K (async)
(anonymous) @ file-uploader.DcK3d__D.js:1
(anonymous) @ file-uploader.DcK3d__D.js:1
tryRun @ file-uploader.DcK3d__D.js:1
(anonymous) @ file-uploader.DcK3d__D.js:1
addTask @ file-uploader.DcK3d__D.js:1
W @ file-uploader.DcK3d__D.js:1
await in W (async)
m @ 2.DdyS7M7E.js:1
v @ 2.DdyS7M7E.js:1
(anonymous) @ scheduler.D4cMnZ5O.js:1
(anonymous) @ scheduler.D4cMnZ5O.js:1
utils.CYnbjJfa.js:1 
        
        
       POST https://<host>/api/assets net::ERR_CONNECTION_RESET
(anonymous) @ utils.CYnbjJfa.js:1
Fr @ utils.CYnbjJfa.js:1
K @ file-uploader.DcK3d__D.js:1
await in K (async)
(anonymous) @ file-uploader.DcK3d__D.js:1
(anonymous) @ file-uploader.DcK3d__D.js:1
tryRun @ file-uploader.DcK3d__D.js:1
(anonymous) @ file-uploader.DcK3d__D.js:1
addTask @ file-uploader.DcK3d__D.js:1
W @ file-uploader.DcK3d__D.js:1
await in W (async)
m @ 2.DdyS7M7E.js:1
v @ 2.DdyS7M7E.js:1
(anonymous) @ scheduler.D4cMnZ5O.js:1
(anonymous) @ scheduler.D4cMnZ5O.js:1
photos:1 Unchecked runtime.lastError: Could not establish connection. Receiving end does not exist.
assets.store.BWy4M-2q.js:1 [Violation] 'setTimeout' handler took 96ms
photos:1 Unchecked runtime.lastError: Could not establish connection. Receiving end does not exist.
```


### Additional information

The Receiving end does not exist errors showed up later, small files (few mb) upload without a hitch.

https://stackoverflow.com/a/73115301 might be useful

</div>

--- a/web/src/lib/utils/file-uploader.ts
+++ b/web/src/lib/utils/file-uploader.ts
@@ -157,7 +157,6 @@ async function fileUploader(
         }
       } catch (error) {
         console.error(`Error calculating sha1 file=${assetFile.name})`, error);
-        throw error;
       }
     }
 

